zerver/worker/queue_processors.py

```diff
@@ -575,9 +568,13 @@ def consume(self, event: Dict[str, Any]) -> None:
 
             # When we process an event, just put it into the queue and ensure we have a timer going.
             user_profile_id = event["user_profile_id"]
-            if user_profile_id not in self.batch_start_by_recipient:
-                self.batch_start_by_recipient[user_profile_id] = time.time()
-            self.events_by_recipient[user_profile_id].append(event)
+
+            batch, newly_created = MissedMessageEmailBatch.objects.get_or_create(
+                user_profile_id=user_profile_id
+            )
+            events = batch.events
+            events.append(event)
+            batch.save(update_fields=["events"])
```

Comment:
We probably want `transaction.atomic` around these modification blocks, so that we don't create an object without appending events if interrupted, for example.

Reply:
OK, done!